### ISSUE TITLE
[FIX-020] Reject nonfinite numeric configuration before creating a run

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#230]
**Category:** Configuration and compatibility · **Priority:** P2 · **Size:** M

**Evidence status:** Reproduced input-validation gaps.

## Outcome and scope

Reject nonfinite public numeric hyperparameters at their configuration boundaries before optimizers, runs or RoPE tables are created. Complete an explicit scalar inventory, preserving valid zero and None meanings. Runtime tensor NaNs, internal ±infinity masks, integer cardinalities and arbitrary application-specific custom config remain outside this repair.

## Current behavior / evidence

`NNOptimParams`, `NNSchedulerParams` and `NNParamGroupSpec` mostly check inequalities, which do not reject NaN. Transformer params do not validate `rope_base`, and direct `RoPE` computes frequencies from any base. The observed NaN-LR configuration is rejected later by installed PyTorch; do not describe that observation as proof of silent invalid optimizer updates. The defect is delayed/inconsistent validation and nonfinite parameter-derived calculations.

## Repository map

- [`src/nnx/nn/params/nn_optim_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_optim_params.py#L51) — Global LR, decay, clipping and momentum/betas.
- [`src/nnx/nn/params/nn_scheduler_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_scheduler_params.py#L31) — Scheduler real-valued fields and existing ranges.
- [`src/nnx/finetune/param_groups.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/finetune/param_groups.py#L61) — Per-group LR/multiplier/decay validation differs from global LR.
- [`src/nnx/nn/params/nn_transformer_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_transformer_params.py#L55) — Positive finite rope_base boundary.
- [`src/nnx/nn/net/transformer_layers.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/net/transformer_layers.py#L47) — Direct construction must enforce base validation too.
- [`src/nnx/peft/lora.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/peft/lora.py#L78) — Adapter scaling constructor boundary; audit analogous DoRA scaling.
- [`tests/test_params_boundary_validation.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_params_boundary_validation.py#L37) — Pattern for named-field errors and legacy valid boundaries.
- [`tests/test_nn_optim_params_builder.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_nn_optim_params_builder.py#L37) — Default omission and direct/builder parity baseline.

## Implementation approach

1. Add a small private finite-real validator, proposed `src/nnx/_validation.py`; use `numbers.Real`, reject booleans/non-real values and then `math.isfinite`. Apply domain checks after finiteness. Do not coerce numeric strings, and avoid rewriting valid historical scalar representations merely to validate them.
2. Record this inventory in tests: global `max_lr`/`weight_decay` ≥0; optional clipping >0; SGD momentum ≥0 (Nesterov >0), Adam beta components in [0,1); group `lr`/`lr_multiplier` >0 and group decay ≥0; scheduler `min_lr`/`threshold` ≥0, factor >0 (plateau <1), optional scheduler `max_lr` >0; RoPE base >0. Preserve probability/dropout domains already enforced. Extend LoRA/DoRA positive alpha guards; direct RMSNorm epsilon must be finite and nonnegative; EarlyStopping.min_delta remains finite and ≥0. Do not turn scheduler integer counts into real-valued fields.
3. Validate direct constructors and params/from-state routes consistently, before model allocation or run creation. Keep meaningful `None` sentinels. Derived per-group LR multiplication must also remain finite; validate resolved values before optimizer construction.
4. Parameterize NaN and both infinities for every inventoried field. Retain valid snapshots and default omission. Reuse a helper if FIX-021 lands first, but do not create a general validation framework or wait for that independent ticket.

## Acceptance criteria

- [ ] Each inventoried field rejects NaN and ±infinity with an error naming the field and accepted domain.
- [ ] `rope_base` rejects zero/negative/nonfinite values both through params and direct RoPE construction; valid configurations produce finite tables.
- [ ] Existing meaningful zeros remain valid, including global LR 0, weight decay 0, and scheduler thresholds 0; existing `None` sentinels remain valid.
- [ ] Invalid values fail before run-directory creation, optimizer mutation, or expensive training; valid state round-trips retain legacy hashes.
- [ ] Parameterized boundary tests include valid near-boundary values and show a normal tiny optimization/Transformer forward remains finite.

## Validation and completion

Minimal existing-API setup:

```python
from dataclasses import replace
import pytest
from nnx import NNOptimParams
from nnx.nn.net.transformer_layers import RoPE

valid = NNOptimParams.builder().adam(max_lr=0.0, weight_decay=0.0).build()
with pytest.raises(ValueError, match="grad_clip_norm"):
    replace(valid, grad_clip_norm=float("nan"))
with pytest.raises(ValueError, match="base"):
    RoPE(dim=4, max_seq_len=4, base=0.0)
```

Add proposed `tests/test_finite_config_validation.py` with a table of constructors/fields/domains. Include base 1.0, LR/threshold zero, nondefault valid beta tuples, None clipping, and finite-overflowing group LR products. Assert errors mention the actual field and domain; assert invalid params create no temporary run directory.

Existing commands: `uv run --frozen pytest tests/test_params_boundary_validation.py tests/test_nn_optim_params_builder.py tests/test_nn_scheduler_params_builder.py tests/test_transformer_layers.py tests/test_finetune_param_groups.py`. After creating the new file run `uv run --frozen pytest tests/test_finite_config_validation.py tests/test_params_round_trip.py`. A tiny valid optimization and Transformer forward must produce finite results; this is not a GPU performance test.

From the repository root, CI-equivalent setup is `uv sync --frozen --all-extras`. The commands above describe required implementation checks; none were executed during ticket enrichment. Finish with `uv run --frozen ruff check src/ tests/ examples/ scripts/`, `uv run --frozen ruff format --check src/ tests/ examples/ scripts/`, and `uv run --frozen pyright --warnings`. Preserve Python 3.10 compatibility and default-omission serialization; update canonical documentation sources, not generated projections by hand.

## Cascading changes and affected consumers

- **Construction and reload boundaries.** Real-valued validation is consumed not only by direct constructors but by [NNTrainParams.from_state](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_train_params.py#L108) and [NNTrainerParams.from_state](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/params.py#L120). Invalid optimizer/scheduler values embedded in a run YAML must fail with their field identity before either training loop allocates an optimizer or claims the run. Valid historical defaults, including zero global LR/decay and None clipping, must keep the same state representation and run ID. Previously accepted NaN/Inf states are invalid inputs to reject, not values to sanitize silently.
- **Resolved values and direct layers.** Group overrides are resolved by [build_param_groups](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/finetune/param_groups.py#L108) in both enum and Trainer construction. Test overflowing finite LR products through these higher APIs, including strict named groups, so helper validation cannot be bypassed after config construction. Direct RoPE, adapter alpha and EarlyStopping constructors remain real public entry points with the same named-field errors; protect them without recasting parameter tensors or integer counts.
- **Samples and user guidance.** The LM sample accepts `--lr` and builds a Transformer plus optimizer; exercise its reusable config-building boundary with NaN LR and a finite control without training the default corpus. Explain finite domains and meaningful zeros in the parameter guide and source docstrings, including distinct global versus per-group LR rules. Keep numeric-string targets and runtime tensor finiteness outside this configuration repair. If AdamW/factory or precision features land later, apply the same validation utility to their declared scalar fields rather than duplicating or prematurely requiring those features.

### Samples and executable coverage

- [`examples/11_tinystories_lm.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/11_tinystories_lm.py): Extract or expose a small config-building helper so the existing CLI configuration can be checked without tokenization/training; do not download the HF corpus in smoke. **Verify:** Proposed test_lm_sample_rejects_nonfinite_config imports the sample, calls its config helper with NaN LR, then a finite control, and asserts a named ValueError versus unchanged valid state.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | State finite scalar domains, global zero-LR allowance, positive group LR and the behavior of invalid serialized configurations. |
| [`docs/lm.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/lm.md) | `generated/site/Language-modeling.md` | `generated/wiki/Language-modeling.md` | Document positive finite rope_base and finite optimizer/dropout settings; distinguish configuration validation from runtime tensor divergence. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] Invalid nonfinite optimizer/scheduler values in nested NNTrainParams and NNTrainerParams states are rejected before a higher-level training call creates a run or optimizer.
- [ ] An overflowing resolved LR multiplier fails through both ordinary NNModel and strict named-Trainer group construction with the offending group identified.
- [ ] The bounded LM sample config check rejects NaN LR and accepts zero global LR without running its corpus or changing valid serialized defaults.
- [ ] The issue-specific changes in `docs/concepts.md`, `docs/lm.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** The traced downstream changes belong in this ticket. Shared-file coordination and optional combinations do not add a hard prerequisite.

Additional nonblocking consumers/coordination: [FEAT-013](https://github.com/thekaveh/NNx/issues/235).

## Delivery order and dependencies

- **Execution order:** 31 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FIX-009](https://github.com/thekaveh/NNx/issues/208), [FIX-021](https://github.com/thekaveh/NNx/issues/231). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. Coordinate validation conventions with FIX-021; does not replace runtime metric FIX-009.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** real-valued hyperparameters were guarded only by inequalities, which NaN passes, so NaN / ±inf (and booleans) reached optimizer construction, run creation or RoPE table generation before anything complained; `NNTransformerParams.rope_base` and a direct `RoPE(base=...)` were not validated at all and silently built non-finite tables for zero / negative / non-finite bases.

**What changed**
- New private `nnx._validation.require_finite_real(value, field, *, owner, minimum, maximum, exclusive_*, domain_message)`: rejects non-real values (booleans and numeric strings are never coerced), then NaN/±inf, then the domain, with a `ValueError` naming the field and its accepted domain; `domain_message` keeps every established user-facing wording. Validation never touches `state()`.
- Wired at every inventoried boundary — direct constructors, `from_state`, `dataclasses.replace` and the trainer-params route alike: `NNOptimParams.max_lr` / `weight_decay` ≥ 0, `grad_clip_norm` > 0 when set, SGD `momentum` ≥ 0, Adam betas in `[0, 1)`; `NNSchedulerParams.factor` > 0, `min_lr` / `threshold` ≥ 0, `max_lr` > 0 when set (the mixed count/real loop is split so the integer counts stay with FIX-021); `NNParamGroupSpec.lr` / `lr_multiplier` > 0 and `weight_decay` ≥ 0, plus the *resolved* per-group learning rate in `build_param_groups`; `NNTransformerParams.rope_base` > 0 and direct `RoPE(base=...)`; `RMSNorm(eps)` finite ≥ 0; `EarlyStopping(min_delta)` finite ≥ 0; LoRA / DoRA `alpha` finite > 0. Meaningful zeros and `None` sentinels stay valid; serialized state and run identity are unchanged.
- Example 11: `build_lm_train_params(lr, ...)` extracted (main uses it) plus the bounded `lm_config_rejects_nonfinite()` check (`tests/test_examples_smoke.py -k lm_config_rejects_nonfinite`, no tokenizer / training / downloads). Docs: `concepts.md` §2.3 "Finite numeric domains", examples catalog, CHANGELOG, API reference.

**PRs:** #297 (develop, merged), #298 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1820 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New `tests/test_finite_config_validation.py`: 19 inventory rows × NaN/+inf/−inf, non-real rejection, meaningful zeros / `None` sentinels with identical `state()`, RoPE base domain and finite tables, `replace` / `from_state` / trainer routes, no run directory before failure, overflowing group-LR product, and a tiny valid optimization plus Transformer forward staying finite. All pre-existing message-matching tests pass unchanged.
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / projection tests) → pass
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #297 and #298.

**Coordination:** integer counts (FIX-021, #231) and dataset batch-size tuples (FIX-022, #232) reuse this validator; runtime tensor NaNs and internal ±inf masks remain outside this repair. Live Pages/wiki publication is left to the `docs.yml` main-push job.
